<a href="https://colab.research.google.com/github/connectalizain/LangChain_HelloWorld/blob/main/RAG_Advance.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
!pip install -q -U google-generativeai

In [3]:
import google.generativeai as genai

In [4]:
from google.colab import userdata
genai.configure(api_key=userdata.get('GOOGLE_API_KEY'))

To find models available:
 and select which model to use:



In [5]:
list(genai.list_models())

[Model(name='models/chat-bison-001',
       base_model_id='',
       version='001',
       display_name='PaLM 2 Chat (Legacy)',
       description='A legacy text-only model optimized for chat conversations',
       input_token_limit=4096,
       output_token_limit=1024,
       supported_generation_methods=['generateMessage', 'countMessageTokens'],
       temperature=0.25,
       max_temperature=None,
       top_p=0.95,
       top_k=40),
 Model(name='models/text-bison-001',
       base_model_id='',
       version='001',
       display_name='PaLM 2 (Legacy)',
       description='A legacy model that understands text and generates text as an output',
       input_token_limit=8196,
       output_token_limit=1024,
       supported_generation_methods=['generateText', 'countTextTokens', 'createTunedTextModel'],
       temperature=0.7,
       max_temperature=None,
       top_p=0.95,
       top_k=40),
 Model(name='models/embedding-gecko-001',
       base_model_id='',
       version='001',
      

EMBEDDING DIRECTLY WITH SDK

In [6]:
from typing import Dict

result: Dict = genai.embed_content(
    model= "models/text-embedding-004",
    content= "My name is Ali Zain",
    task_type= "retrieval_document",
    # optional to add title
    title="embedding single liine"
)
len(result['embedding'])

768

Length of embedded text(Vectors) must be 768


# BUILDING VECTOR STORES and Retrieval using CHROMO DB and LANGCHAIN



In [7]:
!pip install -Uq langchain-chroma

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.3/67.3 kB 4.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 611.1/611.1 kB 14.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 56.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 284.2/284.2 kB 15.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 95.2/95.2 kB 5.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 62.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.6/101.6 kB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.0/16.0 MB 83.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.9/55.9 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 92.0/92.0 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.5/62.5 kB 3.7 MB/s eta 0:00:0

In [8]:
import os
import getpass

In [9]:
from langchain_core.documents import Document

documents = [
    Document(
        page_content="Dogs can understand up to 250 words and gestures.",
        metadata={"animal": "dog", "type": "fact", "source": "canine_facts.txt"}
    ),
    Document(
        page_content="Cats purr not only when they’re happy, but also when they're healing.",
        metadata={"animal": "cat", "type": "health", "source": "feline_behavior.pdf"}
    ),
    Document(
        page_content="Parrots live up to 80 years and need lifelong companionship.",
        metadata={"animal": "parrot", "type": "care", "source": "birdcare_blog"}
    ),
    Document(
        page_content="Hamsters hoard food in their cheeks and store it in hidden spots.",
        metadata={"animal": "hamster", "type": "behavior", "source": "pet_encyclopedia"}
    ),
    Document(
        page_content="Rabbits require hay as a primary part of their diet for digestion.",
        metadata={"animal": "rabbit", "type": "diet", "source": "vet_guidelines"}
    ),
    Document(
        page_content="Goldfish can recognize their owners and be trained to do tricks.",
        metadata={"animal": "goldfish", "type": "fact", "source": "aquarium_notes"}
    )
]


install genai again but with langchain

In [10]:
!pip install -Uq langchain-google-genai

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 2.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 18.2 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-generativeai 0.8.5 requires google-ai-generativelanguage==0.6.15, but you have google-ai-generativelanguage 0.6.17 which is incompatible.


In [11]:
from langchain_google_genai import GoogleGenerativeAIEmbeddings


os.environ["GOOGLE_API_KEY"] = userdata.get('GOOGLE_API_KEY')
embeddings = GoogleGenerativeAIEmbeddings(model="models/text-embedding-004")

In [12]:
len(embeddings.embed_query("I love Pak"))

768

In [13]:
from langchain_chroma import Chroma
# From Langchain_openai impport OpenAiEmbeddings

vectorStore = Chroma.from_documents(
    documents,
    embedding = embeddings
)

In [14]:
list(dir(vectorStore))
# to check functions of vectorstore

['_Chroma__ensure_collection',
 '_Chroma__query_collection',
 '_LANGCHAIN_DEFAULT_COLLECTION_NAME',
 '__abstractmethods__',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__slots__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_asimilarity_search_with_relevance_scores',
 '_chroma_collection',
 '_client',
 '_client_settings',
 '_collection',
 '_collection_metadata',
 '_collection_name',
 '_cosine_relevance_score_fn',
 '_embedding_function',
 '_euclidean_relevance_score_fn',
 '_get_retriever_tags',
 '_max_inner_product_relevance_score_fn',
 '_persist_directory',
 '_select_relevance_score_fn',
 '_similarity_search_with_relevance_scores',
 'aadd_documents',
 'aa

In [15]:
vectorStore.similarity_search("cat")

[Document(id='1b635039-f812-420a-a279-8383d7724d31', metadata={'animal': 'cat', 'source': 'feline_behavior.pdf', 'type': 'health'}, page_content="Cats purr not only when they’re happy, but also when they're healing."),
 Document(id='49e95c19-ed0d-4837-b5f1-413cd7369cfb', metadata={'animal': 'parrot', 'source': 'birdcare_blog', 'type': 'care'}, page_content='Parrots live up to 80 years and need lifelong companionship.'),
 Document(id='98fd10f1-e968-4259-8b38-4431ce6ee0a7', metadata={'animal': 'goldfish', 'source': 'aquarium_notes', 'type': 'fact'}, page_content='Goldfish can recognize their owners and be trained to do tricks.'),
 Document(id='b4ae04b2-843e-4f4d-88a7-afca671046b2', metadata={'animal': 'dog', 'source': 'canine_facts.txt', 'type': 'fact'}, page_content='Dogs can understand up to 250 words and gestures.')]

In [16]:
await vectorStore.asimilarity_search("cat")

[Document(id='1b635039-f812-420a-a279-8383d7724d31', metadata={'animal': 'cat', 'source': 'feline_behavior.pdf', 'type': 'health'}, page_content="Cats purr not only when they’re happy, but also when they're healing."),
 Document(id='49e95c19-ed0d-4837-b5f1-413cd7369cfb', metadata={'animal': 'parrot', 'source': 'birdcare_blog', 'type': 'care'}, page_content='Parrots live up to 80 years and need lifelong companionship.'),
 Document(id='98fd10f1-e968-4259-8b38-4431ce6ee0a7', metadata={'animal': 'goldfish', 'source': 'aquarium_notes', 'type': 'fact'}, page_content='Goldfish can recognize their owners and be trained to do tricks.'),
 Document(id='b4ae04b2-843e-4f4d-88a7-afca671046b2', metadata={'animal': 'dog', 'source': 'canine_facts.txt', 'type': 'fact'}, page_content='Dogs can understand up to 250 words and gestures.')]

In [17]:
vectorStore.similarity_search_with_score("cat")

[(Document(id='1b635039-f812-420a-a279-8383d7724d31', metadata={'animal': 'cat', 'source': 'feline_behavior.pdf', 'type': 'health'}, page_content="Cats purr not only when they’re happy, but also when they're healing."),
  0.6888855695724487),
 (Document(id='49e95c19-ed0d-4837-b5f1-413cd7369cfb', metadata={'animal': 'parrot', 'source': 'birdcare_blog', 'type': 'care'}, page_content='Parrots live up to 80 years and need lifelong companionship.'),
  1.0885951519012451),
 (Document(id='98fd10f1-e968-4259-8b38-4431ce6ee0a7', metadata={'animal': 'goldfish', 'source': 'aquarium_notes', 'type': 'fact'}, page_content='Goldfish can recognize their owners and be trained to do tricks.'),
  1.0914521217346191),
 (Document(id='b4ae04b2-843e-4f4d-88a7-afca671046b2', metadata={'animal': 'dog', 'source': 'canine_facts.txt', 'type': 'fact'}, page_content='Dogs can understand up to 250 words and gestures.'),
  1.1056462526321411)]

# Retriever


In [18]:
from langchain_core.runnables import RunnableLambda

retriever = RunnableLambda(vectorStore.similarity_search).bind(k=1)

retriever.batch(["shark"])

[[Document(id='98fd10f1-e968-4259-8b38-4431ce6ee0a7', metadata={'animal': 'goldfish', 'source': 'aquarium_notes', 'type': 'fact'}, page_content='Goldfish can recognize their owners and be trained to do tricks.')]]

In [25]:
from langchain_google_genai import ChatGoogleGenerativeAI
llm = ChatGoogleGenerativeAI(model="gemini-1.5-flash")
# No need to pass api key again because of os

In [33]:
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables.passthrough import RunnablePassthrough

template = """
Answer this question using only the provided context.

Question: {question}

Context: {context}
"""



In [35]:
promt = ChatPromptTemplate.from_messages(["human", template])

# RAG Pipeline

In [37]:
rag_Chain = {"context": retriever, "question": RunnablePassthrough()} | promt | llm

In [40]:
respose = rag_Chain.invoke("Tell me about animals")
print(respose.content)

Dogs can understand up to 250 words and gestures.


# Face Detection with Embedding

In [42]:
!pip install -qU facenet-pytorch

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 25.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 57.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 58.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 755.6/755.6 MB 695.6 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.6/410.6 MB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.1/14.1 MB 65.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.7/23.7 MB 53.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.6/823.6 kB 34.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 731.7/731.7 MB 821.0 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 121.6/121.6 MB 7.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.5/56.5 MB 11.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 124

In [48]:
!pip install -qU pillow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.6/4.6 MB 28.3 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
facenet-pytorch 2.6.0 requires Pillow<10.3.0,>=10.2.0, but you have pillow 11.2.1 which is incompatible.


In [49]:
import torch
import torch.nn as nn
import torchvision.models as models
import torchvision.transforms as transforms
from PIL import Image
